File: db_orm.go
/*
 * gorm
 */
package main


import (
        "github.com/jinzhu/gorm"
        "encoding/json"
        _ "github.com/jinzhu/gorm/dialects/sqlite"
        _ "fmt"
)

type RequestTable struct {
        gorm.Model
        FirstLine string
        RequestURI string
        Host       string
        SrcIp      string
        SrcPort    string
        DstIp      string
        DstPort    string
        IsResp     int
		StatusCode int
}


type ResponseTable struct{
	gorm.Model
	RequestTable RequestTable `gorm:"foreignkey:RequestRefer"`
	RequestRefer int

    FirstLine    string
	StatusCode   int
    SrcIp        string
    SrcPort      string
    DstIp        string
    DstPort      string
}

type HeaderTable struct{
    gorm.Model
    Type       uint // 1.request header ,2. Response
    Parentid   uint // requestTable or responstTable ID
    Name        string
    Values      string
}

type FormTable struct{
    gorm.Model
    Type       uint // 1.request header ,2. Response
    Parentid   uint // requestTable or responstTable ID
    Name        string
    Values      string
}



type returnReqs struct{
    Reqs   []RequestTable;
    Total  int;
}

func InitDB()(newdb *gorm.DB) {

        newdb, _ = gorm.Open("sqlite3", "./httpdump.db")

		//newdb.LogMode(true)

        newdb.AutoMigrate(&RequestTable{},
            &ResponseTable{},
            &HeaderTable{},
            &FormTable{})

        return newdb
}

func InsertHeaders(db *gorm.DB,n string,v string,t uint,id uint){

    db.Create(&HeaderTable{Type:t,Parentid:id,Name:n,Values:v})
}

func InsertForm(db *gorm.DB,n string,v string,t uint,id uint){

    db.Create(&FormTable{Type:t,Parentid:id,Name:n,Values:v})
}



// insert a request record
func InsertRequestData(db * gorm.DB,req *RequestTable ) (id uint){
    req.IsResp = 1
    db.Create(req)
    return req.ID
}

//update a request record
func UpdateRequestData(db * gorm.DB,req *RequestTable,id uint ){
    db.Model(&RequestTable{}).Where("id=?",id).Updates(req)
}
func UpdateRequestIsResp(db *gorm.DB,req *RequestTable,StatusCode int){

	// IsResp , Column name is_resp
    if req.ID == 0 {
        return
    }
	db.Model(req).Updates(map[string]interface{}{"is_resp":2,"status_code":StatusCode})
}


//insert response record
func InsertResponseData(db *gorm.DB,resp * ResponseTable,req *RequestTable)(id uint){
	// update request status
    // only update IsResp
	UpdateRequestIsResp(db,req,resp.StatusCode)
	//save response
	resp.RequestTable = *req
	db.Create(resp)
    return resp.ID
}

//Update response record
func UpdateResponseData(db *gorm.DB,resp * ResponseTable,req *RequestTable,id uint){
	// update request status again,
    // update StatusCode

	UpdateRequestIsResp(db,req,resp.StatusCode)

	//update response
    db.Model(&ResponseTable{}).Where("id=?",id).Updates(resp)
}

// query all request access record
func FindRequestData(db *gorm.DB) (reqs  []RequestTable){

    db.Order("ID desc").Find(&reqs)

    return reqs
}

// query page data
func FindRequestDataPage(db * gorm.DB,page int, pagesize int) (ret returnReqs){

    var reqs []RequestTable;
    var count int;

	db.Order("ID desc").Limit(pagesize).Offset(page * pagesize).Find(&reqs)

    db.Table("request_tables").Count(&count)

    ret.Reqs  = reqs
    ret.Total = count
	return ret

}

// Get a request id
func FindRequestFirst(db * gorm.DB,SrcIp string,SrcPort string,DstIp string,DstPort string) (req RequestTable){

	db.Where(RequestTable{IsResp:1, SrcIp: DstIp, SrcPort:DstPort,DstIp:SrcIp,DstPort:SrcPort}).First(&req)

	return req
}

// Get request and response information
func FindRequestById(db * gorm.DB,id int)(map[string]interface{}){

    var req RequestTable;
    var resp ResponseTable;
    var reqH,respH []HeaderTable;
    var reqForm []FormTable;

    returnJson := make(map[string]interface{})
    respJson := make(map[string]interface{})
    reqHJson := make([]map[string]interface{},50)
    reqFJson := make([]map[string]interface{},100)
    respHJson := make([]map[string]interface{},50)

    db.First(&req,id)

    // converting to map
    inrec, _ := json.Marshal(&req)
    json.Unmarshal(inrec, &returnJson)

    db.Model(&req).Related(&resp, "RequestRefer")

    // converting to map
    inresp, _ := json.Marshal(&resp)
    json.Unmarshal(inresp, &respJson)

    returnJson["resp"] = respJson

    // get request Header
    db.Where("parentid=? and type=1",req.ID).Find(&reqH)
    // converting to map
    inreqh, _ := json.Marshal(&reqH)
    json.Unmarshal(inreqh, &reqHJson)
    returnJson["reqh"] = reqHJson

    // get request Form 
    db.Where("parentid=? and type=1",req.ID).Find(&reqForm)
    // converting to map
    inreqform, _ := json.Marshal(&reqForm)
    json.Unmarshal(inreqform, &reqFJson)
    returnJson["reqform"] = reqFJson


    // get response Header
    db.Where("parentid=? and type=2",resp.ID).Find(&respH)
    // converting to map
    inresph, _ := json.Marshal(&respH)
    json.Unmarshal(inresph, &respHJson)
    returnJson["resph"] = respHJson


    return returnJson
}

File: db_orm.py
# use SqlAlchemy

from sqlalchemy import create_engine,ForeignKey
from sqlalchemy.orm import sessionmaker,relationship
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import  String,Column,Integer,DateTime
import datetime
engine = create_engine("mysql+mysqldb://cpython:cpython.org@localhost:3306/httpdump")

Session = sessionmaker(bind=engine)
ses = Session()
ses.autocommit == True
Base = declarative_base()

class RequestTable(Base):
    __tablename__ = "request_tables" 
    id = Column(Integer, primary_key=True)
    created_at  = Column(DateTime, default=datetime.datetime.utcnow)
    first_line  = Column(String)
    is_resp     = Column(Integer)
    status_code = Column(Integer)
    request_uri = Column(String)
    host        = Column(String)
    src_ip      = Column(String)
    src_port    = Column(String)
    dst_ip      = Column(String)
    dst_port    = Column(String)
    def serialize(self):
        return {
            'ID': self.id, 
            'CreatedAt': self.created_at,
            'FirstLine': self.first_line,
            'StatusCode':self.status_code,
            'Host':self.host,
            'RequestURI':self.request_uri,
            'SrcIp':self.src_ip,
        }

class ResponseTable(Base):
    __tablename__ = "response_tables"
    id = Column(Integer, primary_key=True)
    created_at    = Column(DateTime, default=datetime.datetime.utcnow)
    request_refer = Column(Integer, ForeignKey("request_tables.id"))  
    first_line    = Column(String)
    status_code   = Column(Integer)
    src_ip      = Column(String)
    src_port    = Column(String)
    dst_ip      = Column(String)
    dst_port    = Column(String)
    
def insert_request(first_line,host,request_uri,src_ip,src_port,dst_ip,dst_port):
    req1 = RequestTable(first_line=first_line,
            host=host,
            request_uri=request_uri,
            src_ip=src_ip,
            src_port=src_port,
            dst_ip=dst_ip,
            dst_port=dst_port,
            is_resp = '0')
    ses.add(req1)
    ses.commit()

def insert_response(src_ip,src_port,dst_ip,dst_port,status_code):
    req1 = ses.query(RequestTable).filter(RequestTable.src_ip==dst_ip,
            RequestTable.src_port==dst_port,
            RequestTable.dst_ip==src_ip,
            RequestTable.dst_port==src_port,
            RequestTable.is_resp=='0').update({'is_resp':'1',
                'status_code':status_code})
    ses.commit()

def get_requests(offset,limit):
    ses1 = Session()
    reqs = ses1.query(RequestTable) \
        .order_by(RequestTable.id.desc())\
        .limit(limit)\
        .offset(offset).all()
    return [r.serialize() for r in reqs]

